### PR TITLE
[SYCLomatic] Emit warning for 3 elements vector types

### PR DESCRIPTION
**Pull request by yihwang-nv**
Signed-off-by: Wang, Yihan <yihan.wang@intel.com>

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/test/dpct/sharedmem-dim-noUSM.cu [hunk 1]
```diff
@@ -83,9 +67,7 @@ __global__ void kernel3() {
 }
 
 void foo3() {
-  // CHECK: /*
-  // CHECK-NEXT: DPCT1083:{{[0-9]+}}: The size of local memory in the migrated code may be different from the original code. Check that the allocated memory size in the migrated code is correct.
-  // CHECK-NEXT: */
-  // CHECK-NEXT: sycl::local_accessor<int, 1> a_acc_ct1(sycl::range<1>(sizeof(sycl::float3) * 3), cgh);
+  // CHECK: DPCT1083:{{[0-9]+}}: The size of local memory in the migrated code may be different from the original code. Check that the allocated memory size in the migrated code is correct.
+  // CHECK: sycl::local_accessor<int, 1> a_acc_ct1(sycl::range<1>(sizeof(sycl::float3) * 3), cgh);
```
Comment (zhimingwang36) [inline clang/test/dpct/sharedmem-dim-noUSM.cu]: Please double check if the changes in this file required.   Only update the test which is impacted by this PR.

File: clang/lib/DPCT/Diagnostics.inc [hunk 2]
```diff
@@ -225,3 +225,5 @@ DEF_WARNING(MODULE_LOAD_DATA, 1104, "'%0' should point to a dynamic library load
 DEF_COMMENT(MODULE_LOAD_DATA, 1104, "'{0}' should point to a dynamic library loaded in memory. The dynamic library should supply \"wrapped\" kernel functions.")
 DEF_WARNING(MODULE_LOAD_DATA_EX, 1105, "'%0' should point to a dynamic library loaded in memory. The dynamic library should supply \"wrapped\" kernel functions. Compilation options passed to cuModuleLoadDataEx are not migrated.")
 DEF_COMMENT(MODULE_LOAD_DATA_EX, 1105, "'{0}' should point to a dynamic library loaded in memory. The dynamic library should supply \"wrapped\" kernel functions. Compilation options passed to cuModuleLoadDataEx are not migrated.")
+DEF_WARNING(SIZEOF_FLOAT3, 1106, "The size of float3 is 12 bytes in CUDA but 16 bytes in SYCL. You may need to adjust the code.")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Diagnostics.inc]:
```suggestion
DEF_WARNING(SIZEOF_FLOAT3, 1106, "The size of sycl::float3 is 16 bytes while in original code size of float3 is 12 bytes. You may need to adjust the code.")
```

File: clang/test/dpct/sharedmem-dim-noUSM.cu (hunk 1 above)
Comment (yihwang-nv) [reply clang/test/dpct/sharedmem-dim-noUSM.cu]: done

File: clang/lib/DPCT/Diagnostics.inc (hunk 2 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/Diagnostics.inc]: > 

done

**Review by zhimingwang36 — APPROVED** [review]
LGTM

File: clang/lib/DPCT/Diagnostics.inc [hunk 3]
```diff
@@ -225,3 +225,5 @@ DEF_WARNING(MODULE_LOAD_DATA, 1104, "'%0' should point to a dynamic library load
 DEF_COMMENT(MODULE_LOAD_DATA, 1104, "'{0}' should point to a dynamic library loaded in memory. The dynamic library should supply \"wrapped\" kernel functions.")
 DEF_WARNING(MODULE_LOAD_DATA_EX, 1105, "'%0' should point to a dynamic library loaded in memory. The dynamic library should supply \"wrapped\" kernel functions. Compilation options passed to cuModuleLoadDataEx are not migrated.")
 DEF_COMMENT(MODULE_LOAD_DATA_EX, 1105, "'{0}' should point to a dynamic library loaded in memory. The dynamic library should supply \"wrapped\" kernel functions. Compilation options passed to cuModuleLoadDataEx are not migrated.")
+DEF_WARNING(SIZEOF_FLOAT3, 1106, "The size of sycl::float3 is 16 bytes while in original code size of float3 is 12 bytes. You may need to adjust the code.")
```
Comment (YuriPlyakhin) [inline clang/lib/DPCT/Diagnostics.inc]: Are those sizes an implementation detail or came from some standard?

File: clang/test/dpct/sharedmem-dim-noUSM.cu [hunk 4]
```diff
@@ -82,10 +82,18 @@ __global__ void kernel3() {
...
 }
 
+// CHECK: void foo3() {
+// CHECK: dpct::get_default_queue().submit(
+// CHECK-NEXT: [&](sycl::handler &cgh) {
+// CHECK: /*
+// CHECK-NEXT: DPCT1083:{{[0-9]+}}: The size of local memory in the migrated code may be different from the original code. Check that the allocated memory size in the migrated code is correct.
+// CHECK-NEXT: */
+// CHECK: /*
+// CHECK-NEXT: DPCT1106:{{[0-9]+}}: The size of sycl::float3 is 16 bytes while in original code size of float3 is 12 bytes. You may need to adjust the code.
```
Comment (YuriPlyakhin) [inline clang/test/dpct/sharedmem-dim-noUSM.cu]: Seems like 2 warnings are about the same: DPCT1083 and DPCT1106. Should they be merged somehow?

**Review by YuriPlyakhin — CHANGES_REQUESTED** [review]

File: clang/lib/DPCT/Diagnostics.inc (hunk 3 above)
Comment (zhimingwang36) [reply clang/lib/DPCT/Diagnostics.inc]: for sizeof(sycl::float3), ref to sycl spec:  https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#memory-layout-and-alignment

File: clang/test/dpct/sharedmem-dim-noUSM.cu (hunk 4 above)
Comment (zhimingwang36) [reply clang/test/dpct/sharedmem-dim-noUSM.cu]: they are same type warnning, while 1106 is specific to "float3",  let's see if can merge.

Comment (yihwang-nv) [reply clang/test/dpct/sharedmem-dim-noUSM.cu]:
```
// CHECK: void test_shared() {
// CHECK: dpct::get_default_queue().submit(
// CHECK-NEXT: [&](sycl::handler &cgh) {
// CHECK: /*
// CHECK-NEXT: DPCT1083:{{[0-9]+}}: The result of sizeof operator in the migrated code may be different from the original code. Check that the size in the migrated code is correct.
// CHECK-NEXT: */
// CHECK-NEXT: sycl::local_accessor<int, 1> a_acc_ct1(sycl::range<1>(sizeof(sycl::float3) * 3), cgh);
// CHECK: });
// CHECK: }
void test_shared() {
  kernel<<<1, 1>>>();
}
```
I would like to merge these two warnings as `The result of sizeof operator in the migrated code may be different from the original code. Check that the size in the migrated code is correct.`, WDYT ?

**Comment by yihwang-nv** [comment]
I have merged DPCT1083 and DPCT1106, also, emit warning for more 3 elements vector types.

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**